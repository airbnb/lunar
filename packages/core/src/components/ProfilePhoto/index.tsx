import React from 'react';
import PropTypes from 'prop-types';
import { and, mutuallyExclusiveProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

const mutuallyExclusiveSizePropType = mutuallyExclusiveProps(
  PropTypes.any,
  'small',
  'large',
  'macro',
  'size',
);
const namedSizePropType = and([PropTypes.bool, mutuallyExclusiveSizePropType]);
const unitSizePropType = and([PropTypes.number, mutuallyExclusiveSizePropType]);

export type Props = {
  /** URL of the image to display. */
  imageSrc: string;
  /** Whether the component should be inline. */
  inline?: boolean;
  /** Large size (96px). */
  large?: boolean;
  /** Macro size (160px). */
  macro?: boolean;
  /** Size in units to multiply by. */
  size?: number;
  /** Small size (24px). */
  small?: boolean;
  /** Whether the image should be square (it is rounded by default). */
  square?: boolean;
  /** Accessibility text for the Photo. */
  title: string;
};

/** Display a profile photo. */
export class ProfilePhoto extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    large: namedSizePropType,
    macro: namedSizePropType,
    size: unitSizePropType,
    small: namedSizePropType,
  };

  // Do not supply defaults for the mutually exclusive props,
  // doing so will cause validation errors.
  static defaultProps = {
    inline: false,
    square: false,
  };

  render() {
    const {
      cx,
      imageSrc,
      inline,
      macro,
      large,
      small,
      size,
      square,
      styles,
      title,
      theme,
    } = this.props;
    const { unit } = theme!;

    return (
      <div
        className={cx(
          inline && styles.inline,
          styles.regular,
          macro && styles.macro,
          large && styles.large,
          small && styles.small,
          !!size &&
            size > 0 && {
              height: size * unit,
              maxHeight: size * unit,
              maxWidth: size * unit,
              width: size * unit,
            },
        )}
      >
        <img
          className={cx(
            styles.image,
            styles.regular,
            !square && styles.roundedImage,
            macro && styles.macro,
            large && styles.large,
            small && styles.small,
            !!size &&
              size > 0 && {
                height: size * unit,
                maxHeight: size * unit,
                maxWidth: size * unit,
                width: size * unit,
              },
          )}
          src={imageSrc}
          alt={title}
          title={title}
        />
      </div>
    );
  }
}

export default withStyles(
  ({ color }) => ({
    inline: {
      display: 'inline-block',
    },
    image: {
      display: 'block',
      background: color.core.neutral[6],
      color: color.base,
      objectFit: 'cover',
      overflow: 'hidden',
    },
    roundedImage: {
      borderRadius: '50%',
    },
    small: {
      height: 24,
      width: 24,
      maxHeight: 24,
      maxWidth: 24,
    },
    regular: {
      height: 48,
      width: 48,
      maxHeight: 48,
      maxWidth: 48,
    },
    large: {
      height: 96,
      width: 96,
      maxHeight: 96,
      maxWidth: 96,
    },
    macro: {
      height: 160,
      width: 160,
      maxHeight: 160,
      maxWidth: 160,
    },
  }),
  {
    passThemeProp: true,
  },
)(ProfilePhoto);
