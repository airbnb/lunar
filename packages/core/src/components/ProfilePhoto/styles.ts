import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color }) => ({
  inline: {
    display: 'inline-block',
  },

  image: {
    display: 'block',
    background: color.core.neutral[6],
    color: color.base,
    objectFit: 'cover',
    overflow: 'hidden',
  },

  roundedImage: {
    borderRadius: '50%',
  },

  small: {
    height: 24,
    width: 24,
    maxHeight: 24,
    maxWidth: 24,
  },
  regular: {
    height: 48,
    width: 48,
    maxHeight: 48,
    maxWidth: 48,
  },

  large: {
    height: 96,
    width: 96,
    maxHeight: 96,
    maxWidth: 96,
  },

  macro: {
    height: 160,
    width: 160,
    maxHeight: 160,
    maxWidth: 160,
  },
});

export default styleSheet;

export { styleSheet };
