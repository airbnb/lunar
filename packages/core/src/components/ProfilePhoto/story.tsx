import React from 'react';
import lunar from ':storybook/images/lunar-logo.png';
import ProfilePhoto from '.';

export default {
  title: 'Core/ProfilePhoto',
  parameters: {
    inspectComponents: [ProfilePhoto],
  },
};

export function defaultProfilePhoto() {
  return <ProfilePhoto imageSrc={lunar} title="Photo" />;
}

defaultProfilePhoto.story = {
  name: 'Default profile photo.',
};

export function withDifferentSizingUnitsBasedSizeSmallRegularDefaultLargeMacro() {
  return (
    <>
      <ProfilePhoto imageSrc={lunar} title="Photo" size={2} />
      <ProfilePhoto small imageSrc={lunar} title="Photo" />
      <ProfilePhoto imageSrc={lunar} title="Photo" />
      <ProfilePhoto large imageSrc={lunar} title="Photo" />
      <ProfilePhoto macro imageSrc={lunar} title="Photo" />
    </>
  );
}

withDifferentSizingUnitsBasedSizeSmallRegularDefaultLargeMacro.story = {
  name: 'With different sizing: units-based size, small, regular (default), large, macro.',
};

export function inlineProfilePhotos() {
  return (
    <>
      <ProfilePhoto inline imageSrc={lunar} title="Photo" size={2} />
      <ProfilePhoto inline small imageSrc={lunar} title="Photo" />
      <ProfilePhoto inline imageSrc={lunar} title="Photo" />
      <ProfilePhoto inline large imageSrc={lunar} title="Photo" />
      <ProfilePhoto inline macro imageSrc={lunar} title="Photo" />
    </>
  );
}

inlineProfilePhotos.story = {
  name: 'Inline profile photos.',
};

export function squareProfilePhoto() {
  return <ProfilePhoto inline large square imageSrc={lunar} title="Photo" />;
}

squareProfilePhoto.story = {
  name: 'Square profile photo.',
};

export function brokenImageWithOverflowingAltText() {
  return <ProfilePhoto large imageSrc="BROKEN_IMAGE" title="Alt title text." />;
}

brokenImageWithOverflowingAltText.story = {
  name: 'Broken image with overflowing alt text.',
};

export function fallbackImageSrcWhenImageIsBroken() {
  return (
    <ProfilePhoto large fallbackImageSrc={lunar} imageSrc="BROKEN_IMAGE" title="Alt title text." />
  );
}

fallbackImageSrcWhenImageIsBroken.story = {
  name: 'Fallback image src when image is broken.',
};
