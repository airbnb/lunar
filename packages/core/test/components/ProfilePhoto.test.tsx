import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import ProfilePhoto from '../../src/components/ProfilePhoto';

describe('<ProfilePhoto />', () => {
  const imageSrc = 'https://domain.com/some/file.jpg';
  const title = 'Name';
  const props = {
    imageSrc,
    title,
  };

  it('renders with the expected props (but not necessarily with the expected styles)', () => {
    const wrapper = shallowWithStyles(<ProfilePhoto {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('correctly sets the image title', () => {
    const wrapper = shallowWithStyles(<ProfilePhoto {...props} />);
    expect(wrapper.find(`img[title="${title}"]`)).toHaveLength(1);
  });

  it('renders inline', () => {
    const wrapper = shallowWithStyles(<ProfilePhoto {...props} inline />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders small', () => {
    const wrapper = shallowWithStyles(<ProfilePhoto {...props} small />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders large', () => {
    const wrapper = shallowWithStyles(<ProfilePhoto {...props} large />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders macro', () => {
    const wrapper = shallowWithStyles(<ProfilePhoto {...props} macro />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders custom size', () => {
    const wrapper = shallowWithStyles(<ProfilePhoto {...props} size={8} />);

    expect(wrapper).toMatchSnapshot();
  });
});
