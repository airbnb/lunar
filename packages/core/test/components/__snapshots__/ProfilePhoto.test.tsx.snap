// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfilePhoto /> renders custom size 1`] = `
<div
  className="regular_i5aehb-o_O-inline-0_x9m9b"
>
  <img
    alt="Name"
    className="image_qfr8ll-o_O-regular_i5aehb-o_O-roundedImage_1bpjyj1-o_O-inline-0_x9m9b"
    src="https://domain.com/some/file.jpg"
    title="Name"
  />
</div>
`;

exports[`<ProfilePhoto /> renders inline 1`] = `
<div
  className="inline_36rlri-o_O-regular_i5aehb"
>
  <img
    alt="Name"
    className="image_qfr8ll-o_O-regular_i5aehb-o_O-roundedImage_1bpjyj1"
    src="https://domain.com/some/file.jpg"
    title="Name"
  />
</div>
`;

exports[`<ProfilePhoto /> renders large 1`] = `
<div
  className="regular_i5aehb-o_O-large_1kuwd2n"
>
  <img
    alt="Name"
    className="image_qfr8ll-o_O-regular_i5aehb-o_O-roundedImage_1bpjyj1-o_O-large_1kuwd2n"
    src="https://domain.com/some/file.jpg"
    title="Name"
  />
</div>
`;

exports[`<ProfilePhoto /> renders macro 1`] = `
<div
  className="regular_i5aehb-o_O-macro_137gz1r"
>
  <img
    alt="Name"
    className="image_qfr8ll-o_O-regular_i5aehb-o_O-roundedImage_1bpjyj1-o_O-macro_137gz1r"
    src="https://domain.com/some/file.jpg"
    title="Name"
  />
</div>
`;

exports[`<ProfilePhoto /> renders small 1`] = `
<div
  className="regular_i5aehb-o_O-small_n3ao3z"
>
  <img
    alt="Name"
    className="image_qfr8ll-o_O-regular_i5aehb-o_O-roundedImage_1bpjyj1-o_O-small_n3ao3z"
    src="https://domain.com/some/file.jpg"
    title="Name"
  />
</div>
`;

exports[`<ProfilePhoto /> renders with the expected props (but not necessarily with the expected styles) 1`] = `
<div
  className="regular_i5aehb"
>
  <img
    alt="Name"
    className="image_qfr8ll-o_O-regular_i5aehb-o_O-roundedImage_1bpjyj1"
    src="https://domain.com/some/file.jpg"
    title="Name"
  />
</div>
`;
